pkg/apis/resource/validation/validation.go

```diff
@@ -191,32 +191,38 @@ func gatherAllocatedDevices(allocationResult *resource.DeviceAllocationResult) s
 
 func validateDeviceRequest(request resource.DeviceRequest, fldPath *field.Path, stored bool) field.ErrorList {
 	allErrs := validateRequestName(request.Name, fldPath.Child("name"))
-
 	numDeviceRequestType := 0
-	if len(request.FirstAvailable) > 0 {
+
+	hasFirstAvailable := len(request.FirstAvailable) > 0
+	hasExactly := request.Exactly != nil
+
+	if hasFirstAvailable {
 		numDeviceRequestType++
-		allErrs = append(allErrs, validateSet(request.FirstAvailable, resource.FirstAvailableDeviceRequestMaxSize,
-			func(subRequest resource.DeviceSubRequest, fldPath *field.Path) field.ErrorList {
-				return validateDeviceSubRequest(subRequest, fldPath, stored)
-			},
-			func(subRequest resource.DeviceSubRequest) (string, string) {
-				return subRequest.Name, "name"
-			},
-			fldPath.Child("firstAvailable"))...)
 	}
-
-	if request.Exactly != nil {
+	if hasExactly {
 		numDeviceRequestType++
-		allErrs = append(allErrs, validateExactDeviceRequest(*request.Exactly, fldPath.Child("exactly"), stored)...)
 	}
 
 	switch numDeviceRequestType {
 	case 0:
 		allErrs = append(allErrs, field.Required(fldPath, "exactly one of `exactly` or `firstAvailable` is required"))
 	case 1:
-	default:
+		if hasFirstAvailable {
```

Comment:
Maybe a bit more readable like this?

```Go
switch {
   case numDeviceRequestType == 0:
       allErrs = append(allErrs, field.Required(fldPath, "exactly one of `exactly` or `firstAvailable` is required"))
   case numDeviceRequestType > 1:
         allErrs = append(allErrs, field.Invalid(fldPath, nil, "exactly one of `exactly` or `firstAvailable` is required, but multiple fields are set"))
   case hasFirstAvailable:
       …
   case hasExactly:
      ...
}
```

Then we don't need the `// hasExactly must be true` comment because it is visible what the branch is for.

Reply:
Ah good point, I've updated the code to the more readable form as suggested